btree/btree.c

```diff
@@ -2902,29 +2947,47 @@ static int64_t btree_put_tick(struct m0_sm_op *smop)
 			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
 		}
 	case P_ALLOC: {
-		int ksize;
-		int vsize;
-		/* Validate curr_level->l_node(i.e.is it still exists or not) */
-		int rc = node_is_valid(curr_level->l_node);
-		if (rc) {
-			node_op_fini(&oi->i_nop);
-			return fail(bop, rc);
-		}
-		if (curr_level->l_node->n_type->nt_id == BNT_FIXED_FORMAT) {
-			ksize = node_keysize(curr_level->l_node);
-			vsize = node_valsize(curr_level->l_node);
-		} else {
-			/**
-			 * @todo : ksize, vsize for variable key:
-			 * get the correct max key size at runtime
-			 * */
-			ksize = MAX_KEY_SIZE;
-			vsize = MAX_VAL_SIZE;
-		}
-		if (m0_btree_overflow_is_possible(curr_level->l_node, ksize,
-						  vsize))
-			return m0_btree_put_alloc_phase(bop);
+		int           ksize;
+		int           vsize;
+		bool          alloc = false;
+		int           rc;
+		do {
+			lev = &oi->i_level[oi->i_used];
+			/* Validate lev->l_node */
+			rc = node_is_valid(lev->l_node);
+			if (rc) {
+				node_op_fini(&oi->i_nop);
+				return fail(bop, rc);
+			}
+			if (lev->l_node->n_type->nt_id == BNT_FIXED_FORMAT) {
+				ksize = node_keysize(lev->l_node);
+				vsize = node_valsize(lev->l_node);
+			} else {
+				/**
+				 * @todo : ksize, vsize for variable key:
+				 * get the correct max key size at runtime
+				 * */
+				ksize = MAX_KEY_SIZE;
+				vsize = MAX_VAL_SIZE;
+			}
+			if (!m0_btree_overflow_is_possible(lev->l_node,
+							   ksize, vsize))
+				break;
+			if (oi->i_used == 0) {
+				if (lev->l_alloc == NULL ||
+				    oi->i_extra_node == NULL)
+					alloc = true;
+				break;
+			} else if (lev->l_alloc == NULL) {
+				alloc = true;
+				break;
```

Comment:
Do we really need a do-while loop. Here we are starting the loop from leaf level. If overflow is possible and l_alloc will always be NULL at the beginning so it will always exit from leaf node. In what condition this loop will traverse upwards ?

Reply:
do-while loop will help, when check failed but height of tree is same, and we start from P_DOWN , so here we can first start checking from leaf , first check if overflow possible, then if allocation is already done i.e. l_alloc != NULL, so, l_alloc will not be always null, it will be null when it is the first traversal but not in every traversal